shared/hexDiffModel.ts

```diff
@@ -0,0 +1,104 @@
+import { bulkhead } from "cockatiel";
+import * as vscode from "vscode";
+import { HexDecorator } from "./decorators";
+import {
+	DiffDecoratorResponseMessage,
+	DiffExtensionHostMessageHandler,
+	DiffMessageType,
+} from "./diffWorkerProtocol";
+import { HexDocumentModel } from "./hexDocumentModel";
+export type HexDiffModelBuilder = typeof HexDiffModel.Builder.prototype;
+
+export class HexDiffModel {
+	/** Guard to make sure only one computation operation happens */
+	private readonly saveGuard = bulkhead(1, Infinity);
+	private decorators?: { original: HexDecorator[]; modified: HexDecorator[] };
+
+	constructor(
+		private readonly originalModel: HexDocumentModel,
+		private readonly modifiedModel: HexDocumentModel,
+		private readonly messageHandler: DiffExtensionHostMessageHandler,
+	) {}
+
+	public async computeDecorators(uri: vscode.Uri): Promise<HexDecorator[]> {
+		return this.saveGuard.execute(async () => {
+			if (this.decorators === undefined) {
+				//TODO: Add a warning if the file sizes are too large?
+				const oSize = await this.originalModel.sizeWithEdits();
+				const mSize = await this.modifiedModel.sizeWithEdits();
+				if (oSize === undefined || mSize === undefined) {
+					throw new Error(vscode.l10n.t("HexEditor Diff: Failed to get file sizes."));
+				}
+
+				const oArray = new Uint8Array(oSize);
+				const mArray = new Uint8Array(mSize);
+				await this.originalModel.readInto(0, oArray);
+				await this.modifiedModel.readInto(0, mArray);
+				const decorators = await this.messageHandler.sendRequest<DiffDecoratorResponseMessage>({
```

Comment:
Both DOM Worker and worker_thread's messaging allow passing an array of [transferrable](https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage#transfer) object to transfer ownership rather than copying memory. Adopting these in the message handler for this case would be a nice to have :)